sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataobject/SnowflakeTableDataObject.scala

```diff
@@ -19,93 +19,85 @@
+  private var _snowparkSession: Option[Session] = None
+
+  def session: Session = {
+    if (_snowparkSession.isEmpty) {
+      _snowparkSession = Some(connection.getSnowparkSession(table.db.get))
+    }
+    _snowparkSession.get
+  }
 
-  // prepare table
   if (table.db.isEmpty) {
     throw ConfigurationException(s"($id) A SnowFlake schema name must be added as the 'db' parameter of a SnowflakeTableDataObject.")
   }
 
-  override def getDataFrame(partitionValues: Seq[PartitionValues] = Seq())(implicit context: ActionPipelineContext): DataFrame = {
+  // Get a Spark DataFrame with the table contents for Spark transformations
+  override def getDataFrame(partitionValues: Seq[PartitionValues] = Seq())(implicit context: ActionPipelineContext): SparkDataFrame = {
     val queryOrTable = Map(table.query.map(q => ("query", q)).getOrElse("dbtable" -> (connection.database + "." + table.fullName)))
     val df = context.sparkSession
       .read
       .format(SNOWFLAKE_SOURCE_NAME)
       .options(connection.getSnowflakeOptions(table.db.get))
       .options(queryOrTable)
       .load()
-    validateSchemaMin(df, "read")
     df.colNamesLowercase
   }
 
-  override def writeDataFrame(df: DataFrame, partitionValues: Seq[PartitionValues] = Seq(), isRecursiveInput: Boolean = false, saveModeOptions: Option[SaveModeOptions] = None)
+  // Write a Spark DataFrame to the Snowflake table
+  override def writeDataFrame(df: SparkDataFrame, partitionValues: Seq[PartitionValues], isRecursiveInput: Boolean, saveModeOptions: Option[SaveModeOptions])
                              (implicit context: ActionPipelineContext): Unit = {
-    validateSchemaMin(df, "write")
-    writeDataFrame(df, createTableOnly = false, partitionValues, saveModeOptions)
-  }
-
-  /**
-   * Writes DataFrame to Snowflake
-   * Snowflake does not support explicit partitions, so any passed partition values are ignored
-   */
-  def writeDataFrame(df: DataFrame, createTableOnly: Boolean, partitionValues: Seq[PartitionValues], saveModeOptions: Option[SaveModeOptions])
-                    (implicit context: ActionPipelineContext): Unit = {
-    implicit val session: SparkSession = context.sparkSession
-    val dfPrepared = if (createTableOnly) {
-      DataFrameUtil.getEmptyDataFrame(df.schema)
-    } else {
-      df
-    }
+    validateSchemaMin(df, role = "write")
+    val finalSaveMode: SDLSaveMode = saveModeOptions.map(_.saveMode).getOrElse(saveMode)
 
-    val finalSaveMode = saveModeOptions.map(_.saveMode).getOrElse(saveMode)
-    dfPrepared.write
+    df.write
       .format(SNOWFLAKE_SOURCE_NAME)
       .options(connection.getSnowflakeOptions(table.db.get))
       .options(Map("dbtable" -> (connection.database + "." + table.fullName)))
       .mode(finalSaveMode.asSparkSaveMode)
       .save()
 
-    if (comment.isDefined) {
-      val sql = s"comment on table ${connection.database}.${table.fullName} is '${comment.get}';"
+    if (comment != null && comment.isDefined) {
```

Comment:
It should not happen that an Option is null. I see that the default value is missing in the definition as case class attribute. Can you fix the default value and remove the null testing here?

Reply:
Fixed. We had some trouble compiling the project due to generic values in the case classes extending ParsableFromConfig, so we did some experiments with removing generics. Forgot to revert this.